# IFN509 Assignment 2

In [28]:
from pathlib import Path
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [29]:
data_file = "Dataset.csv"
df = pd.read_csv(data_file)

## Task 1

Let's take a look at a header of the data

In [30]:
df.head()

Participant_ID survey_date region country  ip_latitude  ip_longitude  \
0       862217754  02/12/2020      ?      US      42.3690      -88.0651   
1         2363592  02/07/2020     AF      ZA     -26.2870       28.2427   
2         2570172  05/07/2020     AF      ZA     -26.2019       28.0273   
3         2812446  13/06/2020     AF      ZA     -33.8268       18.4772   
4         2867850  29/06/2020     AF      ZA     -26.3120       28.1024   

   gender    age  height  weight  ...  heart_disease lung_disease diabetes  \
0    male  30_40     180      98  ...              0            0        0   
1    male  30_40     168      66  ...              0            0        1   
2  female  60_70     172      86  ...              0            0        0   
3    male  40_50     172     140  ...              1            0        1   
4    male  20_30     186      78  ...              0            0        0   

  hiv_positive other_chronic  nursing_home  health_worker  risk_infection  \
0            0             0             0              0           100.0   
1            0             0             0              0            75.0   
2            0             0             0              0             5.0   
3            0             0             0              0           100.0   
4            0             0             0              0             5.0   

   risk_infection_level risk_mortality  
0                  high          0.063  
1              moderate          0.098  
2                   low          0.582  
3                  high          1.032  
4                   low          0.050  

[5 rows x 38 columns]

We'll use .info() to examine the datatypes and .isnull() to examine null values - we'll process these later

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5789 entries, 0 to 5788
Data columns (total 38 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Participant_ID          5789 non-null   int64  
 1   survey_date             5789 non-null   object 
 2   region                  5774 non-null   object 
 3   country                 5786 non-null   object 
 4   ip_latitude             5789 non-null   float64
 5   ip_longitude            5789 non-null   float64
 6   gender                  5789 non-null   object 
 7   age                     5789 non-null   object 
 8   height                  5789 non-null   int64  
 9   weight                  5789 non-null   int64  
 10  bmi                     5789 non-null   float64
 11  blood_type              5689 non-null   object 
 12  insurance               4497 non-null   object 
 13  income                  4636 non-null   object 
 14  smoking                 5753 non-null   

In [32]:
df.isnull().sum()

Participant_ID               0
survey_date                  0
region                      15
country                      3
ip_latitude                  0
ip_longitude                 0
gender                       0
age                          0
height                       0
weight                       0
bmi                          0
blood_type                 100
insurance                 1292
income                    1153
smoking                     36
alcohol                      0
cocaine                   4359
contacts_count              42
public_transport_count    1217
working                     28
worried                   1271
covid19_positive             0
covid19_symptoms             0
covid19_contact              0
asthma                       0
kidney_disease               0
liver_disease                0
compromised_immune           0
heart_disease                0
lung_disease                 0
diabetes                     0
hiv_positive                 0
other_ch

# Column Data Types & Missing Data Review

| #  | Column                  | Non-Null Count | Current Dtype       | Suggested Dtype         | Missing Values / Notes |
|----|-------------------------|----------------|-------------------|------------------------|-----------------------|
| 0  | Participant_ID          | 5789           | int64             | int64                  | 0, fine |
| 1  | survey_date             | 4874           | datetime64[ns]    | datetime64[ns]         | 915 missing; may have malformed dates |
| 2  | region                  | 5774           | category          | category               | 15 missing |
| 3  | country                 | 5786           | category          | category               | 3 missing |
| 4  | ip_latitude             | 5789           | float64           | float64                | 0 missing, fine |
| 5  | ip_longitude            | 5789           | float64           | float64                | 0 missing, fine |
| 6  | gender                  | 5789           | category          | category               | 0 missing, fine |
| 7  | age                     | 5783           | category          | ordered category       | 6 missing; stored as ranges |
| 8  | height                  | 5789           | int64             | int64                  | 0 missing, fine |
| 9  | weight                  | 5789           | int64             | int64                  | 0 missing, fine |
| 10 | bmi                     | 5789           | float64           | float64                | 0 missing, fine |
| 11 | blood_type              | 5689           | category          | category               | 100 missing |
| 12 | insurance               | 4497           | category          | category               | 1292 missing |
| 13 | income                  | 4636           | category          | category               | 1153 missing |
| 14 | smoking                 | 5753           | category          | category               | 36 missing |
| 15 | alcohol                 | 5789           | float64           | float64                | 0 missing, fine |
| 16 | cocaine                 | 1430           | float64           | float64                | 4359 missing |
| 17 | contacts_count          | 5747           | Int64             | Int64 (nullable)       | 42 missing |
| 18 | public_transport_count  | 4572           | Int64             | Int64 (nullable)       | 1217 missing |
| 19 | working                 | 5761           | category          | category               | 28 missing |
| 20 | worried                 | 4518           | float64           | float64                | 1271 missing |
| 21 | covid19_positive        | 5789           | boolean           | boolean                | 0 missing |
| 22 | covid19_symptoms        | 5789           | boolean           | boolean                | 0 missing |
| 23 | covid19_contact         | 5789           | boolean           | boolean                | 0 missing |
| 24 | asthma                  | 5789           | boolean           | boolean                | 0 missing |
| 25 | kidney_disease          | 5789           | boolean           | boolean                | 0 missing |
| 26 | liver_disease           | 5789           | boolean           | boolean                | 0 missing |
| 27 | compromised_immune      | 5789           | boolean           | boolean                | 0 missing |
| 28 | heart_disease           | 5789           | boolean           | boolean                | 0 missing |
| 29 | lung_disease            | 5789           | boolean           | boolean                | 0 missing |
| 30 | diabetes                | 5789           | boolean           | boolean                | 0 missing |
| 31 | hiv_positive            | 5789           | boolean           | boolean                | 0 missing |
| 32 | other_chronic           | 5789           | boolean           | boolean                | 0 missing |
| 33 | nursing_home            | 5789           | boolean           | boolean                | 0 missing |
| 34 | health_worker           | 5789           | boolean           | boolean                | 0 missing |
| 35 | risk_infection          | 5789           | float64           | float64                | 0 missing |
| 36 | risk_infection_level    | 5789           | category          | category               | 0 missing |
| 37 | risk_mortality          | 5789           | float64           | float64                | 0 missing |


Let's convert the data types explicitly, we could use .asType() to do this in one snippet, but it might fail if there's missing or malformed data. Better to do each type explicitly

In [33]:
# --------------------------
# 1. Boolean columns (safe)
# --------------------------
bool_cols = [
    'covid19_positive', 'covid19_symptoms', 'covid19_contact',
    'asthma', 'kidney_disease', 'liver_disease', 'compromised_immune',
    'heart_disease', 'lung_disease', 'diabetes', 'hiv_positive',
    'other_chronic', 'nursing_home', 'health_worker'
]
df[bool_cols] = df[bool_cols].astype('boolean')

# --------------------------
# 2. Safe categorical columns
# --------------------------
safe_cat_cols = [
    'region', 'country', 'gender', 'smoking', 'working', 'risk_infection_level'
]
df[safe_cat_cols] = df[safe_cat_cols].astype('category')

# --------------------------
# 3. Numeric columns that are already clean
# --------------------------
numeric_cols = [
    'ip_latitude', 'ip_longitude', 'height', 'weight', 'bmi',
    'alcohol', 'risk_infection', 'risk_mortality'
]
df[numeric_cols] = df[numeric_cols].astype('float64')

# --------------------------
# 4. Columns to preprocess later (commented)
# --------------------------
# survey_date -> convert to datetime after fixing bad/missing values
# df['survey_date'] = pd.to_datetime(df['survey_date'], errors='coerce')

# age -> ordered category
# age_order = ['0_10','10_20','20_30','30_40','40_50','50_60','60_70','70_80','80_90','90_100']
# df['age'] = pd.Categorical(df['age'], categories=age_order, ordered=True)

# blood_type, insurance, income, cocaine, contacts_count, public_transport_count, worried
# -> preprocess missing / malformed values before conversion


In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5789 entries, 0 to 5788
Data columns (total 38 columns):
 #   Column                  Non-Null Count  Dtype   
---  ------                  --------------  -----   
 0   Participant_ID          5789 non-null   int64   
 1   survey_date             5789 non-null   object  
 2   region                  5774 non-null   category
 3   country                 5786 non-null   category
 4   ip_latitude             5789 non-null   float64 
 5   ip_longitude            5789 non-null   float64 
 6   gender                  5789 non-null   category
 7   age                     5789 non-null   object  
 8   height                  5789 non-null   float64 
 9   weight                  5789 non-null   float64 
 10  bmi                     5789 non-null   float64 
 11  blood_type              5689 non-null   object  
 12  insurance               4497 non-null   object  
 13  income                  4636 non-null   object  
 14  smoking                 

## Task 2